analyzer/merger-cache-invalidation-plugin/src/main/java/eu/fasten/analyzer/mergercacheinvalidationplugin/MergerCacheInvalidationPlugin.java

```diff
@@ -0,0 +1,188 @@
+         */
+        public void loadGraphResolver(String depGraphPath) {
+            logger.info("Building Dependency Graph from " + depGraphPath + "...");
+            try {
+                var graphResolver = new GraphMavenResolver();
+                graphResolver.buildDependencyGraph(null, depGraphPath);
+                this.graphResolver = graphResolver;
+            } catch (Exception e) {
+                logger.error("Couldn't build the dependency graph", e);
+                System.exit(1);
+            }
+            logger.info("...Dependency Graph has been successfully built.");
+        }
+
+        @Override
+        public String name() {
+            return "Merger Cache Invalidation Plugin";
+        }
+
+        @Override
+        public String description() {
+            return "Merger Cache Invalidation Plugin. "
+                    + "Consumes list of updated product from Kafka"
+                    + " topic and invalidates cache of vulnerable paths"
+                    + " for all its transitive dependants.";
+        }
+
+        @Override
+        public String version() {
+            return "0.0.1";
+        }
+
+        @Override
+        public void start() { }
+
+        @Override
+        public void stop() {
+            this.graphResolver = null;
+        }
+
+        public void setPluginError(Exception throwable) {
+            this.pluginError = throwable;
+        }
+
+        @Override
+        public Exception getPluginError() {
+            return this.pluginError;
+        }
+
+        @Override
+        public void freeResource() { }
+
+        @Override
+        public Optional<List<String>> consumeTopic() {
+            return Optional.of(Collections.singletonList(consumerTopic));
+        }
+
+        @Override
+        public void setTopic(String topicName) {
+            this.consumerTopic = topicName;
+        }
+
+        @Override
+        public void consume(String record) {
+            this.pluginError = null;
+
+            // Parse JSON object from kafka topic of GraphDBExtension.
+            // Although it doesn't have output payload, the plugin serializes the graph for its input.
+            // And we can use the input copy from this topic and the serialized graph to process our caching.
+            var json = new JSONObject(record);
+            if (json.has("input")) {
+                if (json.get("input").toString().isEmpty()) {
+                    logger.error("Empty input");
+                    setPluginError(new RuntimeException("Empty input"));
+                    return;
+                }
+                json = json.getJSONObject("input");
+            }
+
+            // Parse input values for the root product.
+            String groupId = "";
+            String artifactId = "";
+            String version = "";
+            try {
+                if (!json.has("product")) {
+                    throw new JSONException("product");
```

Comment:
Does `setPluginError` crashes the plugin as well (e.g., on line 117). Shall I replace it as well?

In general, I was referencing `[MetadataDBExtension.java](https://github.com/fasten-project/fasten/blob/2711be588779b30fce8af8daf5e4600fd6454e9e/analyzer/metadata-plugin/src/main/java/eu/fasten/analyzer/metadataplugin/MetadataDBExtension.java#L113)`. Shall I create an issue for it as well? Or it has a different scenario?

Reply:
> Does `setPluginError` crashes the plugin as well (e.g., on line 117). Shall I replace it as well?
> 
> In general, I was referencing [`MetadataDBExtension.java#L113`](https://github.com/fasten-project/fasten/blob/2711be588779b30fce8af8daf5e4600fd6454e9e/analyzer/metadata-plugin/src/main/java/eu/fasten/analyzer/metadataplugin/MetadataDBExtension.java#L113). Shall I create an issue for it as well? Or it has a different scenario?

In the metadata plugin, that exception on line 113 is watched if you look carefully. No exception should go from the plugin, because this will crash it. When you want to throw an exception, you just use `setPluginError(yourException)` and `return` from the method.

Reply:
In my case, it's also watched, but see it isn't returned. Shall I just return then in the catch clause?

Reply:
I guess I can also reset `depSet` to null in produce after it's converted to json. Wdyt?

Reply:
I think you just need to do `setPluginError(new JsonException(...)); return;`.
And on line 53 do `setPluginError(e); return;`

Reply:
> I think you just need to do setPluginError(new JsonException(...)); return;

I kinda understand this one

> And on line 53 do setPluginError(e); return;

But what do you mean here?

Reply:
> But what do you mean here? You mean the catch clause for graph resolver? Here I'd say we need to crash/stop the plugin, because, without a graph resolver, there is no point in continuing.

No, the plugin should never crash, because then k8s will need to restart it which results in a pretty bad performance.  By doing `setPluginError` and `return` you kind of skip the record and it goes to the `fasten.<plugin-name>.err` topic instead of `fasten.<plugin-name>.out`, and the plugin proceeds to process the next record